osmoutils/binary_search.go

```diff
@@ -0,0 +1,95 @@
+package osmoutils
+
+import (
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ErrTolerance is used to define a compare function, which checks if two
+// ints are within a certain error tolerance of one another.
+// ErrTolerance.Compare(a, b) returns true iff:
+// |a - b| <= AdditiveTolerance
+// |a - b| / min(a, b) <= MultiplicativeTolerance
+// Each check is respectively ignored if the entry is nil (sdk.Dec{}, sdk.Int{})
+type ErrTolerance struct {
+	AdditiveTolerance       sdk.Int
+	MultiplicativeTolerance sdk.Dec
+}
+
+// IsSatisfied returns if actual is within errTolerance of expected.
+// returns 0 if it is
+// returns 1 if not, and expected > actual.
+// returns -1 if not, and expected < actual
+func (e ErrTolerance) Compare(expected sdk.Int, actual sdk.Int) int8 {
+	diff := expected.Sub(actual).Abs()
+
+	comparisonSign := int8(0)
+	if expected.GT(actual) {
+		comparisonSign = 1
+	} else {
+		comparisonSign = -1
+	}
+
+	// if no error accepted, do a direct compare.
+	if e.AdditiveTolerance.IsZero() {
+		if expected.Equal(actual) {
+			return 0
+		} else {
+			return comparisonSign
+		}
+	}
+
+	// Check additive tolerance equations
+	if !e.AdditiveTolerance.IsNil() && !e.AdditiveTolerance.IsZero() {
```

Comment:
Should we check `e.AdditiveTolerance.IsNil()` before `e.AdditiveTolerance.IsZero()` above?

Reply:
Also, `!e.AdditiveTolerance.IsZero()` is always true at this point so it can be removed